### PR TITLE
add IDE and nuget workaround in C# quickstart

### PR DESCRIPTION
**Pull request by apolcyn**
The pending change in https://github.com/grpc/grpc/pull/7623 would break IDE usage on Xamarin Studio and Monodevelop IDEs, so this adds some workarounds, and another story in general by using the .NET Core SDK.

This documentation shouldn't be merged before https://github.com/grpc/grpc/pull/7634 is merged, as it talks about that example.

### CODE REVIEW

File: docs/quickstarts/csharp-quickstart.md [hunk 1]
```diff
@@ -16,9 +16,14 @@ type: markdown
...
 ### Prerequisites
 
-* Windows: .NET Framework 4.5+, Visual Studio 2013 or 2015. You may also need the Nuget executable 
-* Linux: Mono 4+, MonoDevelop 5.9+ (with NuGet add-in installed)
-* Mac OS X: Xamarin Studio 5.9+
+Windows: 
+
+* .NET Framework 4.5+, Visual Studio 2013 or 2015.
+
+For Mac OS X and Linus, either:
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: s/Linus/Linux/

File: docs/quickstarts/csharp-quickstart.md [hunk 2]
```diff
@@ -16,9 +16,14 @@ type: markdown
...
-* Linux: Mono 4+, MonoDevelop 5.9+ (with NuGet add-in installed)
-* Mac OS X: Xamarin Studio 5.9+
+Windows: 
+
+* .NET Framework 4.5+, Visual Studio 2013 or 2015.
+
+For Mac OS X and Linus, either: 
+
+* Mono 4+, MonoDevelop 5.9+, and a nuget executable 
+* The .NET Core sdk, the .NET framework 4.5, and a nuget executable
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: s/sdk/SDK/

File: docs/quickstarts/csharp-quickstart.md [hunk 3]
```diff
@@ -16,9 +16,14 @@ type: markdown
...
-* Windows: .NET Framework 4.5+, Visual Studio 2013 or 2015. You may also need the Nuget executable 
-* Linux: Mono 4+, MonoDevelop 5.9+ (with NuGet add-in installed)
-* Mac OS X: Xamarin Studio 5.9+
+Windows: 
+
+* .NET Framework 4.5+, Visual Studio 2013 or 2015.
+
+For Mac OS X and Linus, either: 
+
+* Mono 4+, MonoDevelop 5.9+, and a nuget executable
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: s/nuget/NuGet/. Here and elsewhere.

File: docs/quickstarts/csharp-quickstart.md [hunk 4]
```diff
@@ -30,18 +35,47 @@ $ # Clone the repository to get the example code:
...
+From the `examples/csharp/helloworld-from-cli` directory:
+
+```
+> dotnet restore
+> dotnet build **/project.json
+```
+
+### Xamarin Studio and Monodevelop
+The C# gRPC package currently depends on System.Interactive.Async 3.0.0. 
+If you are using the Xamarin Studio or Monodevelop IDEs on mac or linux, this poses an issue because
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: s/mac/OS X/
s/linux/Linux/

Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: This is a bit wordy. How about:

> The C# gRPC package currently depends on System.Interactive.Async 3.0.0, which cannot be downloaded on older NuGet versions. NuGet is too old in Xamarin Studio and Monodevelop on OS X and Linux as well as via apt-get.

File: docs/quickstarts/csharp-quickstart.md [hunk 5]
```diff
@@ -30,18 +35,47 @@ $ # Clone the repository to get the example code:
...
+their nuget clients are currently not at a late enough version to download this nuget dependency.
+The nuget executable available via apt-get currently also has the same problem.
+
+One possible workaround is as follows:
+
+* Upgrade your nuget installation with `/path-to-your-nuget-executable/nuget update -self`.
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: I tend to see this written as `/path/to/nuget`.

File: docs/quickstarts/csharp-quickstart.md [hunk 6]
```diff
@@ -30,18 +35,47 @@ $ # Clone the repository to get the example code:
...
 dependencies for you (Grpc, Grpc.Tools and Google.Protobuf NuGet packages).
 
 ## Build the example
 
-* Open the solution `Greeter.sln` with Visual Studio, Monodevelop (on Linux) or Xamarin Studio (on Mac OS X)
+### Visual Studio
+* Open the solution `Greeter.sln` with Visual Studio.
 * Build the solution (this will automatically download NuGet dependencies)
+
+### Using .NET Core sdk from the command line.
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: Remove the period, since it's a header.

**Comment by apolcyn** [comment]
The most recent commits should address the comments, and a few more typos I caught. This was for the wrong folder and filename though too, so changed that here too.

**Comment by ejona86** [comment]
:-/. The rename and the modifications were in the same commit, so I can't easily see the changes.

This PR is now conflicting with the GA branch.

**Comment by ejona86** [comment]
Yeah... this is a bit unreviewable now, because I would have to review the entire file.

**Comment by apolcyn** [comment]
would squashing since the first review help here?

File: docs/quickstarts/csharp-quickstart.md [hunk 7]
```diff
@@ -30,35 +35,75 @@ $ # Clone the repository to get the example code:
...
+One possible workaround is as follows:
+
+* Upgrade your NuGet installation with `/path/to/nuget update -self`. 
+* From the `examples/csharp/helloworld` directory, run `/path/to/nuget restore`. 
+* Now that the NuGet dependencies are restored into their proper package folders, build
+  the solution from your IDE.
   
 ## Run a gRPC application
 
+### After building with Visual Studio, Xamarin Studio, or Monodevelop IDEs:
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: This and the rest of the headers aren't seeming like headers so much. If you're putting a colon at the end of a header, it probably isn't a header, or it needs to be reworded.

So, for example, here you could have:

```
### Using Visual Studio, Xamarin Studio, or Monodevelop IDEs

After building, from the ...
```

File: docs/quickstarts/csharp-quickstart.md [hunk 8]
```diff
@@ -118,8 +163,22 @@ message HelloReply {
 ## Generate gRPC code
 
 Next we need to update the gRPC code used by our application to use the new service definition. 
-The Grpc.Tools NuGet package contains the protoc and protobuf C# plugin binaries you will need to generate the code. This example project already depends on the NuGet package `Grpc.Tools.1.0.0`, so it should be included in the `examples/csharp/helloworld/packages` when the `Greeter.sln` solution is built from your IDE. 
 
+The Grpc.Tools NuGet package contains the protoc and protobuf C# plugin binaries you will need to generate the code. 
+
+#### If you're following the Visual Studio story or restored packages with NuGet from the command line:
```
Comment (ejona86) [inline docs/quickstarts/csharp-quickstart.md]: This and the next header are the only usages of "story". Let's not use that word, since we aren't using it elsewhere. We should probably just have a "Using blah" header here and below instead.

**Comment by ejona86** [comment]
Pretty close. Just need header tweaking.

The PR still conflicts with GA. You should merge with GA to fix it.

**Comment by apolcyn** [comment]
sorry still haven't fixed conflicts with GA by the way, will need to move the file name and folder too before merging

**Comment by apolcyn** [comment]
rebased, but will need to move the file before merging. this also needs to wait on the other pending PRs as well.

**Comment by ejona86** [comment]
Still a bit hard to read what the changes were, but LGTM.

**Comment by apolcyn** [comment]
I see that the diffs are hard to read. I can squash the last two, or all commits here again if that would help

**Comment by hsaliak** [comment]
@apolcyn ping on the merge conflicts..

**Comment by apolcyn** [comment]
this still needs to wait on the dependency changing PR, but just rebased and fixed remaining merge conflicts

**Comment by hsaliak** [comment]
LGTM feel free to commit once unblocked.

**Comment by apolcyn** [comment]
FYI this is still blocked now blocked on the new cli example added in https://github.com/grpc/grpc/pull/7634. I've ran through this example on osx and linux and seen it work. I'm not sure if we want to run through this later though before possibly merging it in.

**Comment by apolcyn** [comment]
the latest commit should fix the errors found during the quickstart walkthrough. Tried these changes on windows and saw it work. 

cc @nathanielmanistaatgoogle

**Comment by apolcyn** [comment]
The PR https://github.com/grpc/grpc/pull/7634 is ready to merge now, so this should be ready to go.

@nathanielmanistaatgoogle if you could take a look at the change in the last commit please.

**Comment by nathanielmanistaatgoogle** [comment]
5c426d1fb24000557dd7ee7af1984c56a8bce480 luhguhtum.
